sw/device/lib/dif/dif_padctrl.c

```diff
@@ -0,0 +1,305 @@
+    attr_change_type_t change_type) {
+  // Check arguments before doing anything.
+  if (padctrl == NULL) {
+    return kDifPadctrlChangeBadArg;
+  }
+
+  if (!is_pad_index_valid(kind, index)) {
+    return kDifPadctrlChangeBadArg;
+  }
+
+  if (!attrs_are_valid(attrs)) {
+    return kDifPadctrlChangeBadArg;
+  }
+
+  // Cannot continue with locked padctrl as registers cannot be altered.
+  if (padctrl_is_locked(padctrl)) {
+    return kDifPadctrlChangeLocked;
+  }
+
+  // Calculate which register and which field within it relate to the pad that
+  // is the target of the attribute change.
+  bitfield_field32_t attr_field = calc_pad_attr_field(kind, index);
+  uint32_t attr_reg_offset = calc_pad_attr_reg_offset(kind, index);
+
+  // Read current attribute register and extract relevant field.
+  uint32_t attr_reg_value =
+      mmio_region_read32(padctrl->base_addr, attr_reg_offset);
+
+  attr_field.value = bitfield_get_field32(attr_reg_value, attr_field);
+
+  // Alter field depending upon change type.
+  switch (change_type) {
+    case kAttrChangeEnable:
+      attr_field.value |= attrs;
+      break;
+    case kAttrChangeDisable:
+      attr_field.value &= ~attrs;
+      break;
+    case kAttrChangeWrite:
+      attr_field.value = attrs;
+      break;
+    default:
+      return kDifPadctrlBadArg;
+  }
+
+  // Check if the new setting causes a conflict.
+  if (conflicting_attributes(attr_field.value)) {
+    return kDifPadctrlChangeAttrConflict;
+  }
+
+  // Write back updated value to attribute register.
+  uint32_t new_attr_reg_value =
+      bitfield_set_field32(attr_reg_value, attr_field);
+
+  mmio_region_write32(padctrl->base_addr, attr_reg_offset, new_attr_reg_value);
+
+  return kDifPadctrlChangeOk;
+}
+
+dif_padctrl_result_t dif_padctrl_init(mmio_region_t base_addr,
+                                      dif_padctrl_t *padctrl) {
+  if (padctrl == NULL) {
+    return kDifPadctrlBadArg;
+  }
+
+  padctrl->base_addr = base_addr;
+
+  return kDifPadctrlOk;
+}
+
+dif_padctrl_reset_result_t dif_padctrl_reset(dif_padctrl_t *padctrl,
+                                             dif_padctrl_attr_t reset_attrs) {
+  if (padctrl == NULL) {
+    return kDifPadctrlResetBadArg;
+  }
+
+  if (!attrs_are_valid(reset_attrs)) {
+    return kDifPadctrlResetBadArg;
+  }
+
+  if (padctrl_is_locked(padctrl)) {
+    return kDifPadctrlResetLocked;
+  }
+
+  // Form attribute register reset value from reptitions of reset_attrs
+  // appropriately shifted and combined.
+  uint32_t attr_reg_reset_val = 0;
+
+  for (int i = 0; i < PADCTRL_MIO_PADS_ATTR_FIELDS_PER_REG; ++i) {
+    attr_reg_reset_val <<= PADCTRL_PARAM_ATTRDW;
+    attr_reg_reset_val |= reset_attrs;
+  }
+
+  // Reset all DIO pad attributes to the desired reset value.
+  for (int i = 0; i < PADCTRL_DIO_PADS_MULTIREG_COUNT; ++i) {
+    ptrdiff_t dio_pads_register_offset =
+        PADCTRL_DIO_PADS0_REG_OFFSET + (i * kPadctrlRegisterWidthInBytes);
```

Comment:
Why isn't this `calc_pad_attr_reg_offset()`?

Reply:
They aren't quite the same, `calc_pad_attr_reg_offset()` takes a pad kind and index, here i is not a pad index but a register index. I could for instance iterate through pad indexes with a sufficient increment that I only get one pad per register and the use `calc_pad_attr_reg_offset()` but I felt this way is clearer